File: js-sdk/cortex-mem-js/src/dto/extraction.ts
// ============================================================
// Extraction DTOs
// ============================================================

import { safeString, safeStringOr, safeNumberOr, safeRecord } from './wire-helpers';

/**
 * Extraction result from the backend.
 * Used by both latest and history endpoints.
 *
 * Wire format uses camelCase (backend Map.of).
 * Use {@link parseExtractionResult} to safely parse from wire format.
 */
export interface ExtractionResult {
  status?: string;
  template?: string;
  message?: string;
  sessionId: string;
  extractedData: Record<string, unknown>;
  createdAt: number;
  observationId: string;
}

/**
 * Parse a raw wire-format extraction result into the canonical type.
 * Handles null values and type mismatches gracefully.
 */
export function parseExtractionResult(raw: Record<string, unknown>): ExtractionResult {
  return {
    status: safeString(raw.status),
    template: safeString(raw.template),
    message: safeString(raw.message),
    sessionId: safeStringOr(raw.sessionId, ''),
    extractedData: safeRecord(raw.extractedData) ?? {},
    createdAt: safeNumberOr(raw.createdAt, 0),
    observationId: safeStringOr(raw.observationId, ''),
  };
}

File: go-sdk/cortex-mem-go/dto/extraction.go
package dto

// ExtractionResult contains a single extraction result.
// Used by both GET /api/extraction/{template}/latest and GET /api/extraction/{template}/history.
type ExtractionResult struct {
	Status        string         `json:"status,omitempty"`   // "ok" or "not_found" (latest only)
	Template      string         `json:"template,omitempty"` // template name (latest only)
	Message       string         `json:"message,omitempty"`  // error/not-found message (latest only)
	SessionID     string         `json:"sessionId"`
	ExtractedData map[string]any `json:"extractedData"`
	CreatedAt     int64          `json:"createdAt"` // epoch millis
	ObservationID string         `json:"observationId"`
}
